plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/exporter/LocalIndexExporter.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.insights.core.exporter;
+
+import org.opensearch.action.bulk.BulkRequestBuilder;
+import org.opensearch.action.index.IndexRequest;
+import org.opensearch.action.support.WriteRequest;
+import org.opensearch.client.Client;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.common.xcontent.XContentFactory;
+import org.opensearch.core.xcontent.ToXContent;
+import org.opensearch.plugin.insights.rules.model.SearchQueryRecord;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.format.DateTimeFormatter;
+
+import java.util.List;
+
+/**
+ * Local index exporter for exporting query insights data to local OpenSearch indices.
+ */
+public final class LocalIndexExporter extends AbstractExporter {
+    final private Client client;
+    final private DateTimeFormatter indexPattern;
+
+    /**
+     * Constructor of LocalIndexExporter
+     *
+     * @param client OS client
+     * @param indexPattern the pattern of index to export to
+     */
+    public LocalIndexExporter(final Client client, final DateTimeFormatter indexPattern) {
+        this.indexPattern = indexPattern;
+        this.client = client;
+    }
+
+    /**
+     * Export a list of SearchQueryRecord to a local index
+     *
+     * @param records list of {@link SearchQueryRecord}
+     * @return True if export succeed, false otherwise
+     */
+    @Override
+    public boolean export(final List<SearchQueryRecord> records) {
+        if (records == null || records.size() == 0) {
+            return true;
+        }
+        try {
+            final String index = getDateTimeFromFormat();
+            final BulkRequestBuilder bulkRequestBuilder = client.prepareBulk()
+                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+                .setTimeout(TimeValue.timeValueMinutes(1));
+            for (SearchQueryRecord record : records) {
+                bulkRequestBuilder.add(
+                    new IndexRequest(index).source(record.toXContent(XContentFactory.jsonBuilder(), ToXContent.EMPTY_PARAMS))
+                );
+            }
+            bulkRequestBuilder.execute().actionGet();
```

Comment:
Which threadpool this logic is running on? This will block the thread and generally not advisable. Response should ideally be handled using request/response listener callback

Reply:
We have an executor for query insights: https://github.com/opensearch-project/OpenSearch/blob/2a37bddefe876a91ace2faf1a3cc91ebe0b7b6e5/plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java#L166

Let me evaluate the latency impact here, we can change it to request/response listener callback if this is a concern.

Reply:
Thanks, saw the executor later! I still believe it is better to have callbacks instead of blocking the threads on index write

Reply:
Added a callback function to log any errors encountered during bulk.